### ISSUE TITLE
Documentation?

### ISSUE THREAD

**Issue by AAverin** [Adobe-CEP/CEP-Resources#138]
I am reading through https://github.com/Adobe-CEP/CEP-Resources/blob/master/CEP_8.x/Documentation/CEP%208.0%20HTML%20Extension%20Cookbook.md already 3rd time and it still totally unclear where to start, what to install and how to write a panel extension.

Documentation should be better structured and clear for new developers to ensure quick start.

**Comment by mBeierl** [comment]
Hey, I'd recommend checking out one of the [examples](https://github.com/Adobe-CEP/Samples)

**Comment by AAverin** [comment]
Yea, I have found http://www.adobe.com/devnet/creativesuite/articles/a-short-guide-to-HTML5-extensions.html
It explains things a little, but I am still not successful running anything – my simple Hello World just doesn't show up in Window -> Extensions even though I have fixed all the versions in manifest already

**Comment by undavide** [comment]
Hi @AAverin, you can check the [HTML Panels Tips Series](http://www.davidebarranca.com/category/code/html-panels/) on my blog (free).
If you need a more comprehensive approach, I've authored a full course on HTML Panels for sale [here](http://htmlpanelsbook.com/).
HTH,

Davide

**Comment by AAverin** [comment]
@undavide I have them opened in one of the tabs, but aren't those articles a bit outdated?
Also, the first one starts with Debugging – that would be step 3 or 4 for me, because I didn't even run Hello World yet)

**Comment by undavide** [comment]
CEP itself is a bit outdated  If you think that ExtendScript is stuck at the Javascript syntax of 2009... :-D
Just kidding (not really), the basics haven't changed from when I wrote those articles. 
I'm afraid to advertise my own work here, but from the HTML Panels Course website you can get an excerpt of about 60 pages from my book (which is 300 pages in total). Those 60 pages cover Hello World and much more in excruciating details.

**Comment by AAverin** [comment]
@undavide 
Thanks for the advice

btw, do I have to restart Photoshop every time I make changes to the plugin, or it would update automatically?

**Comment by undavide** [comment]
If you make changes to the JSX, you need to close/reopen the Panel, if you change the JS only, a simple panel refresh is OK.
You refresh the panel with Command+R or Ctrl+R in the Chrome Developer Tool, or more usually you add something like this: `<button onclick="window.location.reload()">Refresh</button>` to the panel

**Comment by mds-charu** [comment]
@undavide 

HI, Is this refresh works for Illustrator and Indesign as well?

**Comment by undavide** [comment]
I've never built panels for ID or IL myself, sorry, but I would hardly think it's a PS exclusive feature.
Mind you, if you have enabled "Panel Persistence" – in Photoshop it's disabled by default – *then* you need to restart the host application to reload everything (this is what persistence is about).
Perhaps in IL and ID it's enabled by default? I'm afraid I don't know.

**Comment by mds-charu** [comment]
Ok Thanks, Also I have more question is Adobe plugin development support angular2?

**Comment by undavide** [comment]
Definitely, you can use whatever JS framework you prefer. Please note that HTML Panels are hosted in the Google Chromium Embedded Framework (CEF) and not Google Chrome, so don't just test in the browser but in the actual Panel.

I'm not into TypeScript ES6 Webpack etc. so I have no idea how to set up such things in the context of Extensions development, sorry

**Comment by mds-charu** [comment]
Thanks undavide. If you find any plugin development with angular2. Please share.

**Comment by mBeierl** [comment]
@mds-charu you can set-up your Angular project as you're used to and just point to your app's entry point (e.g. the index.html of your built Angular app) in the manifest.xml. If your build is located in build/index.html, you could e.g. use: 
```
.
.
.
<DispatchInfoList>
		<Extension Id="your.extension.id">
			<DispatchInfo>
				<Resources>
					<MainPath>./build/index.html</MainPath>
.
.
.
```

**Comment by mds-charu** [comment]
I set up the project but when I debug in Indesign, It shows blank panel. It does not show the content of the component. Here is the dropbox link for a sample. If you can have look and let me know what I am missing that will be a great help.

https://www.dropbox.com/s/biv0k5va8bradji/Sample1.zip?dl=0

**Comment by undavide** [comment]
Do you mean the Chrome Developer tool is blank (in case, try downloading Google Chromium instead), or the Panel itself in InDesign?

**Comment by mds-charu** [comment]
Panel itself is blank.  And when I debug in chromium using port localhost:7000(as it was specified in debug file). HTML elements looks like attached screen
![screen shot 2017-10-26 at 4 07 34 pm](https://user-images.githubusercontent.com/6419329/32048642-d222a208-ba67-11e7-81e3-cb34d8b97fa6.png)

**Comment by mBeierl** [comment]
@mds-charu your sample project is missing a manifest.xml file. Here is an example how your manifest.xml should look like: https://github.com/Adobe-CEP/CEP-Resources/blob/master/CEP_8.x/Samples/CEP_HTML_Test_Extension-8.0/CSXS/manifest.xml

It also needs to be in the root/CSXS folder

**Comment by mds-charu** [comment]
Sorry, I have attached my project code not compiled code.
Here is the link for compiled code and it has manifest file. I just coped this extension from CEP folder.

https://www.dropbox.com/s/as8qm3xn4b3r9km/com.example.Sample1.zip?dl=0

**Comment by mBeierl** [comment]
You need to build your project using the `npm run build` command you already set-up in your `my-sample/package.json` and also in your manifest.xml change from:

```
...
<MainPath>./my-sample/src/index.html</MainPath>
<ScriptPath>./Sample1.jsx</ScriptPath>
...
```

to:

```
...
<MainPath>./my-sample/dist/index.html</MainPath>
<ScriptPath>./Sample1.jsx</ScriptPath>
...
```

Additionally you will need to actually use the CSInterface.js that you currently have placed in /lib folder. Rename it to CSInterface.js and put in in your /my-sample/src folder and also add it to the .angular-cli.json file in the assets array, otherwise the Angular CLI won't copy it to your dist folder.

Your .angular.cli.json's apps->assets array should look like this:
```
...
  "apps": [
    {
      "root": "src",
      "outDir": "dist",
      "assets": [
        "assets",
        "favicon.ico",
        "CSInterface.js"
      ],
      "index": "index.html",
...
```

**Comment by mBeierl** [comment]
btw: in the future it would be easier if you just create a quick repo with your project instead of using Dropbox. 

Then I could simply create a PR with the changes needed.

**Comment by AAverin** [comment]
So far the biggest question from me – how to debug jsx files or, at least, see error messages if something goes wrong there?
I already had many cases of some syntax errors that just break everything and are very difficult to find

**Comment by bbb999** [comment]
Here's a (somewhat-boring-but-potentially) useful video I made, showing how to debug PPro panels at both the JavaScript and ExtendScript levels.  [Download before viewing; Dropbox playback is terrible.]

https://www.dropbox.com/s/lwo8jg0klxkq91s/walkthru.mp4

**Comment by AAverin** [comment]
Thanks, looks like I need to check this ExtendScript Toolkit thing

**Comment by bbb999** [comment]
ESTK = the only way to debug ExtendScript.

**Comment by undavide** [comment]
Hi @bbb999, thanks for sharing the video! I'm on PS and not AE, so I watched with surprise the fact the Chrome Dev Tool could jump to ESTK for debugging an `evalScript` call, never seen anything like that in Photoshop. 
I've just tried and it seems the JSX call is simply evaluated – should I set debug points in the ESTK as well, or it's an AE exclusive feat? Thanks!

**Comment by AAverin** [comment]
So far in Photoshop breakpoints are just not hit.
I am able to walk throught the script manually, but if am calling an eval function that supposed to trigger a breakpoint – it just doesn't work for me

**Comment by bbb999** [comment]
Yes, set breakpoints in ESTK before calling, and set the target application to [target application], before attempting to debug. 

That's PPro, not AE, btw. :)

On occasion, I've had to 'jiggle the handle' by hitting Play on the script first.

**Comment by undavide** [comment]
PPro, AE... all software that deals with moving stuff, it's a sin! I only worship still images :-D 
I tried as you suggested in PS, but no way. Have you ever tried yourself with Photoshop?
It would be a great feature request, though! Thanks for sharing anyway.

**Comment by bbb999** [comment]
STILL images?! Aww, that's adorable! ;)

Sadly, I can't get ESTK breakpoints to work in Photoshop, either.

**Comment by AAverin** [comment]
Guys, what about error codes descriptions?
Or some kind of lifecycle? I want to close my custom log files when extension is closed

**Comment by bbb999** [comment]
What's your question/request, around error code descriptions? and lifecycle? 

In PPro, panels get called before they're unloaded; I'm not sure what Photoshop does.

https://github.com/Adobe-CEP/Samples/blob/master/PProPanel/index.html#L334

**Comment by AAverin** [comment]
@bbb999 It's my second day with Adobe scripting and I am mostly busy banging my head on the table. So I have tones of questions and my main request is to have good documentation for this whole thing =)
But speaking more specific:
- I am getting some errors in logs once in a while, like `ERROR AsyncEvalRawScript: Failed to evaluate script for scripting engine <right.apps.arcpanel_Engine_Id> with error code 38`
What is code 38? What other codes there are? Is there a documentation for that?
- I have written a custom logging solution to get at least some info from JSX in Photoshop:
```
function log(message) {
	var now = new Date();
	var output = now.toTimeString() + ": " + message;
        var logFile = File(".../log.txt");
        logFile.open("e");
	logFile.writeln(output);
	$.writeln(output);
}

function suspend(name, action) {
	app.activeDocument.suspendHistory(name, "try { " + action + " } catch(e) { log(e + \" while running " + action + " \" ) }");
}
```
Problem with it now is that file is re-opened every time I want to log anything.
I would like to delete and open new file when extension loads and close it when extension is closed.
Thus the question – is there some kind of lifecycle? Where can I find documentation for it? Is it a regular webpage lifecycle (I just got that thought while typing and probably it's indeed like that)

**Comment by bbb999** [comment]
I have no idea about that error code, or Photoshop scripting. Launching it for the first time in a while this morning, testing for Davide, above, stirred up a cloud of dust. ;)

Broadly, a panel 'is a' web page, with commensurate life cycle.

Have you been in touch with anyone, from the Photoshop team?

Even more broadly...what sort of panel are you creating?

**Comment by AAverin** [comment]
@bbb999 I'm making a panel that is kind of like a runner for Actions, except Actions are written by directly calling Photoshop functions. So HTML/CSS wise panel will be very simple and most of the work is done in JSX layer calling various Photoshop methods and filters to provide needed result for the user

How can I reach Photoshop team?
All links about Photoshop scripting and CEP point to this Github repo, so here I am, in the Issues =)

**Comment by bbb999** [comment]
Write me at my day job: b b b at adobe dot com.

**Comment by Sindrom132** [comment]
![default](https://user-images.githubusercontent.com/19823017/44625134-4d830780-a90a-11e8-9dbb-39e326a7a63d.JPG)

I made a mistake. corrected.
As you can see in the picture, the variable res must contain the values ​​obtained from ExtenScript, and so it is but only inside the block...
![default](https://user-images.githubusercontent.com/19823017/44625157-d306b780-a90a-11e8-83ef-6c980c7aa87c.JPG) 
Beyond the borders of this block it shows undefined.

**Comment by rondreas** [comment]
Current documentation is horrible. One should not have to jump between a dozen of samples and god knows how many pdfs it was decided upon to split the whole thing into.

Can't for the life of me find any information regarding error codes or the function "system()" for example.

**Comment by Inventsable** [comment]
That's not specific to CEP. Why would it be in CEP documentation?

**Comment by rondreas** [comment]
I assume the function "system()" is CEP seeing it's not recognized in ExtendScript editor nor any js editor and called upon using csInterface.evalScript or it won't run at all. And I see no structured documentation to find anything regarding this function.

**Comment by bbb999** [comment]
http://docs.aenhancers.com/general/system/?highlight=system

**Comment by rondreas** [comment]
Can't be the same object, system("p4 info > " file.fsName); dumps perforce info in this case, system.userName returns undefined.

**Comment by schroef** [comment]
> Can't be the same object, system("p4 info > " file.fsName); dumps perforce info in this case, system.userName returns undefined. (edit: doing this in Photoshop)

You can check the csinterface-X.js file to see what function you can call. For photoshop you can use this for instance, var OSVersion = csInterface.getOSInformation();

That code system() seems to be a AE internal scripting function. Its not related to CEP in general i think.
